tensorwaves/interfaces.py

```diff
@@ -0,0 +1,102 @@
+"""Defines top level interfaces of tensorwaves.
+
+"""
+
+from abc import ABC, abstractmethod
+
+
+class Function(ABC):
+    """
+    Defines the interface of a callable function.
+    The parameters of the model are separated from the domain variables. This
+    follows the mathematical definition, in which a function defines its domain
+    and parameters. However specific points in the domain are not relevant.
+    Hence while the domain variables are the argument of the evaluation
+    (see :func:`~Function.__call__`), the parameters are controlled via a
+    getter and setter (see :func:`~Function.parameters`). The reason for this
+    separation is to facilitate the events when parameters have changed.
+    """
+    @abstractmethod
+    def __call__(self, dataset: dict) -> list:
+        """Evaluate the function.
+
+        Args:
+            dataset: a dictionary with domain variable names as keys and a
+            an list or `numpy.array` of  with a
+
+        Return:
+            list or :class:`numpy.array` of values
+        """
+
+    @property
+    @abstractmethod
+    def parameters(self) -> dict:
+        """
+        """
+
+    @parameters.setter
+    @abstractmethod
+    def parameters(self, new_parameters: dict):
+        """
+        """
+
+
+class Estimator(ABC):
+    @abstractmethod
+    def __call__(self) -> float:
+        pass
+
+    @property
+    @abstractmethod
+    def parameters(self) -> dict:
+        pass
+
+    @parameters.setter
+    @abstractmethod
+    def parameters(self, new_parameters: dict):
+        pass
+
+
+class Kinematics(ABC):
+    @abstractmethod
+    def convert(self, events: dict) -> dict:
+        pass
+
+    @abstractmethod
+    def reduce_to_phase_space(self, events: dict) -> dict:
+        pass
+
+    @property
+    @abstractmethod
+    def phase_space_volume(self) -> float:
+        pass
+
+
+class Optimizer(ABC):
```

Comment:
An optimizer has to optimize a model. Should some reference to a model be implemented in this abstract interface?

Reply:
The optimizer get an Estimator and a set of parameters. He has to tweak the parameters so that the Estimator return value becomes as small as possible

Reply:
Then I suppose this comment resolves this https://github.com/ComPWA/tensorwaves/pull/30/files/a2d859cf1dfd332e6e18f4f97a2bad5faea54911#r413062124